source/distributed/grid_refinement.cc

```diff
@@ -264,12 +264,13 @@ namespace internal
                                   return c > test_threshold;
                                 });
 
-                unsigned int total_count = 0;
+                // Potentially accumulate in a 64bit int to avoid overflow:
+                types::global_dof_index total_count = 0;
 
                 ierr = MPI_Reduce(&my_count,
                                   &total_count,
                                   1,
-                                  MPI_UNSIGNED,
+                                  DEAL_II_DOF_INDEX_MPI_TYPE,
```

Comment:
We could have avoided having to spell it out by using `MPI::sum` here, which auto-deducts the data type.

Reply:
Yes, I don't like that sum is an allreduce, which costs twice as much. That's why I did not make the change immediately. In general it is probably worth it to avoid bugs like this.

Reply:
Fair point.

Reply:
> Yes, I don't like that sum is an allreduce, which costs twice as much.

I think the most common allreduce implementations for small sizes (like a single int) do not cost twice as much as a reduce because they are not implemented by reduce + bcast. Instead, something called "recursive doubling" is, to the best of my knowledge, the most common algorithm. If reduce takes log2(p) steps (everything going to root, p is a power of 2), then the recursive doubling takes log2(p)+1 steps. The trick is to let all processes take part of the summation; this sends considerably more messages than reduce+bcast but in the end it is the latency that matters.

Reply:
> they are not implemented by reduce + bcast. Instead, something called "recursive doubling" is, to the best of my knowledge, the most common algorithm.

I made a visualization of that ages ago: https://github.com/peterrum/in2012-ws3-allreduce#algorithm